pandas/core/dtypes/cast.py

```diff
@@ -462,6 +501,281 @@ def maybe_promote(dtype, fill_value=np.nan):
+                "boolean": bool,
+                "complex": complex,
+                "bytes": bytes,
+                "datetime64": "datetime64[ns]",
+                "datetime": "datetime64[ns]",
+                "date": "datetime64[ns]",
+                "timedelta64": "timedelta64[ns]",
+                "timedelta": "timedelta64[ns]",
+                "time": object,  # time cannot be cast to datetime/timedelta
+                "string": object,
+                "mixed-integer": object,
+                "mixed": object,
+            }
+            fill_dtype = np.dtype(map_inferred_to_numpy[inferred_dtype])
+
+    # now that we have the correct dtype; check how we must upcast
+    # * extension arrays
+    # * int vs int
+    # * int vs float / complex
+    # * float vs float
+    # * float vs complex (and vice versa)
+    # * bool
+    # * datetimetz
+    # * datetime
+    # * timedelta
+    # * string/object
+
+    # if (is_extension_array_dtype(dtype)
+    #         or is_extension_array_dtype(fill_dtype)):
+    #     # TODO: dispatch to ExtensionDType.maybe_promote? GH 24246
+    if is_integer_dtype(dtype) and is_integer_dtype(fill_dtype):
+        if is_unsigned_integer_dtype(dtype) and all(fill_value >= 0):
+            # can stay unsigned
+            fill_max = fill_value.max()
+            if fill_max > _uint64_max:
+                return np.dtype(object), np.nan
+
+            while fill_max > np.iinfo(dtype).max:
+                # itemsize is the number of bytes; times eight is number of
+                # bits, which is used in the string identifier of the dtype;
+                # if fill_max is above the max for that dtype,
+                # we double the number of bytes/bits.
+                dtype = np.dtype("uint{}".format(dtype.itemsize * 8 * 2))
+            return dtype, None
+        else:
+            # cannot stay unsigned
+            if dtype == "uint64":
+                # need to hold negative values, but int64 cannot hold
+                # maximum of uint64 -> needs object
+                return np.dtype(object), np.nan
+            elif is_unsigned_integer_dtype(dtype):
+                # need to turn into signed integers to hold negative values
+                # int8 cannot hold maximum of uint8; similar for 16/32
+                # therefore, upcast at least to next higher int-type
+                dtype = np.dtype("int{}".format(dtype.itemsize * 8 * 2))
+
+            fill_max = fill_value.max()
+            fill_min = fill_value.min()
+            if isinstance(fill_max, np.uint64):
+                # numpy comparator is broken for uint64;
+                # see https://github.com/numpy/numpy/issues/12525
+                # use .item to get int object
+                fill_max = fill_max.item()
+
+            # comparison mechanics are broken above _int64_max;
+            # use greater equal instead of equal
+            if fill_max >= _int64_max + 1 or fill_min <= _int64_min - 1:
```

Comment:
can you use the can_cast machinery machinery currently in the scalar function?  or even just dispatch to the scalar function in some cases?

Reply:
Dispatching to the scalar case is out of the question for performance reasons until this whole code is cythonized (or the logic somehow unified with `lib.maybe_conver_object`).